### PR TITLE
feat!: add app version to param store

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Ported over from https://github.com/celestiaorg/cosmos-sdk/pull/347

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by evan-forbes — APPROVED** [review]
to double check my understanding, this is a breaking change ofc but it will not be breaking until the app version if > 2, correct?

**Comment by cmwaters** [comment]
> to double check my understanding, this is a breaking change ofc but it will not be breaking until the app version if > 2, correct?

This is the intention. I will be able to test it properly on the celestia-app side

**Review by evan-forbes — APPROVED** [review]

File: baseapp/abci.go [hunk 1]
```diff
@@ -122,7 +125,15 @@ func (app *BaseApp) SetOption(req abci.RequestSetOption) (res abci.ResponseSetOp
...
 	lastCommitID := app.cms.LastCommitID()
-
+	// load the app version for a non zero height
+	if lastCommitID.Version > 0 {
+		ctx, err := app.createQueryContext(lastCommitID.Version, false)
+		if err != nil {
+			panic(err)
+		}
+		// get and set the app version
+		_ = app.AppVersion(ctx)
```
Comment (rootulp) [inline baseapp/abci.go:135]: [question] where or how does this "set" the app version? The invoked function returns the app version but I don't know what "set" is referring to here.

https://github.com/celestiaorg/cosmos-sdk/blob/4c897c6e15c7c0b956ca9fa25d0625ff7a39553a/baseapp/baseapp.go#L208-L211

**Review by evan-forbes — APPROVED** [review]
re-approving even tho we don't have to in this repo to indicate that I have reviewed the fix

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -98,6 +97,13 @@ type BaseApp struct { // nolint: maligned
 	// ResponseCommit.RetainHeight.
 	minRetainBlocks uint64
 
+	// application's version string
+	version string
+
+	// application's protocol version that increments on every upgrade
+	// if BaseApp is passed to the upgrade keeper's NewKeeper method.
+	appVersion uint64
```
Comment (rootulp) [inline baseapp/baseapp.go:105]: [not blocking] can these two include an example to help further differentiate them?

If `version = 2.0.0` then is `appVersion = 2`? Is appVersion always expected to be the major version of the version string?

**Review by rootulp — APPROVED** [review]

**Comment by rootulp** [comment]
On re-review, I think it would be helpful if we updated the README w/ a line item for the rationale for this modification

**Pull request merged**